control-plane-operator/controllers/hostedcontrolplane/nodebootstrapper/kubeconfig.go

```diff
@@ -0,0 +1,195 @@
+		if tokenSecret.Annotations == nil {
+			tokenSecret.Annotations = map[string]string{}
+		}
+		tokenSecret.Annotations["kubernetes.io/service-account.name"] = sa.Name
+		tokenSecret.Type = corev1.SecretTypeServiceAccountToken
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to reconcile bootstrapper service account token secret: %w", err)
+	}
+
+	// Resolve the kubeconfig secret for bootstrapper sa token
+	err = hostedClusterClient.Get(ctx, client.ObjectKeyFromObject(tokenSecret), tokenSecret)
+	if err != nil {
+		return fmt.Errorf("failed to get bootstrapper service account token secret %q: %w", tokenSecret.Name, err)
+	}
+	if tokenSecret.Data[corev1.ServiceAccountRootCAKey] == nil || tokenSecret.Data[corev1.ServiceAccountTokenKey] == nil {
+		return fmt.Errorf("failed to get bootstrapper service account token from secret %q: %w", tokenSecret.Name, err)
+	}
+	if hcp.Status.ControlPlaneEndpoint.Host == "" {
+		return fmt.Errorf("failed to get apiserver url from hcp.Status.ControlPlaneEndpoint: %v", hcp.Status.ControlPlaneEndpoint.Host)
+	}
+	apiServerURL := fmt.Sprintf("https://%s:%d", hcp.Status.ControlPlaneEndpoint.Host, hcp.Status.ControlPlaneEndpoint.Port)
+	// generate kubeconfig from the token secret
+	kcData, _, err := kubeconfigFromSecret(tokenSecret, apiServerURL)
+	if err != nil {
+		return fmt.Errorf("failed to generate kubeconfig from secret %q: %w", tokenSecret.Name, err)
+	}
+
+	bootstrapKubeconfigSecret := manifests.BootstrapKubeconfigSecret(hcp.Namespace)
+	_, err = controllerutil.CreateOrUpdate(ctx, mangementClient, bootstrapKubeconfigSecret, func() error {
+		return reconcileBootstrapKubeconfigSecret(bootstrapKubeconfigSecret, kcData, KubeconfigKey)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to reconcile bootstrapper kubeconfig %q: %w", bootstrapKubeconfigSecret.Name, err)
+	}
+	return nil
+}
+
+func reconcileBootstrapClusterRoleBinding(binding *rbacv1.ClusterRoleBinding, role *rbacv1.ClusterRole, sa *corev1.ServiceAccount) error {
+	binding.RoleRef = rbacv1.RoleRef{
+		APIGroup: "rbac.authorization.k8s.io",
+		Kind:     "ClusterRole",
+		Name:     role.Name,
+	}
+
+	binding.Subjects = []rbacv1.Subject{
+		{
+			Kind:      "ServiceAccount",
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		},
+	}
+	return nil
+}
+
+func reconcileBootstrapKubeconfigSecret(secret *corev1.Secret, kcData []byte, key string) error {
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+	if key == "" {
+		key = "kubeconfig"
+	}
+	if secret.Labels == nil {
+		secret.Labels = map[string]string{}
+	}
+	//secret.Labels[kas.KubeconfigScopeLabel] = string(scope)
+	secret.Labels["hypershift.openshift.io/kubeconfig"] = "bootstrap"
+	secret.Data[key] = kcData
+	return nil
+}
+
+func kubeconfigFromSecret(secret *corev1.Secret, apiserverURL string) ([]byte, []byte, error) {
+	caData := secret.Data[corev1.ServiceAccountRootCAKey]
```

Comment:
I used the ca from service account directly (same as the machine-config-operator), but I guess I can also get ca from root-ca secret? not sure which is better